### ISSUE TITLE
KNX messages duplicated after bridge runs for some time

### ISSUE THREAD

**Issue by giggo** [pakerfeldt/knx-mqtt-bridge#6]
After some time the bridge starts duplicating KNX message (and sending duplicated MQTT messages:
```

verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 2/1/14, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 2/1/14, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 2/1/14, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 2/1/14, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 2/1/14, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 2/1/14, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 2/1/14, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
verbose: 2019-01-31 22:42:43 **** KNX EVENT: GroupValue_Write, dst: 4/2/6, value: "false"
```

My config.yaml is:
```
# One of 'error', 'warn', 'info', 'verbose', debug', 'silly'
loglevel: 'debug'
# One of value-only, full
#    value-only - converts any known group addresses to its value
#    full - a json object containing value. Additionally also name and unit type for known group addresses.
#messageType: value-only
messageType: value-only
# Ignore unknown group addresses
ignoreUnknownGroupAddresses: false
knx:
    # ETS exported group addresses
    # etsExport: '/home/pi/knx-mqtt-bridge/knx_master.xml'
    etsExport: '/home/pi/knx-mqtt-bridge/gruppenadressenets5modified.xml'
    #etsExport: '/home/pi/knx-mqtt-bridge/sterntaler.knxproj'
    # Configuration passed to the KNX library
    options:
    ipAddr: '192.168.1.13'
    ipPort: 3671
    forceTunneling: true
    physAddr: '2.1.11'
    minimumDelay: 200
    suppress_ack_ldatareq: true

mqtt:
    # URL to MQTT broker
    url: 'mqtt://localhost'
    # Configuration passed to the MQTT library
    options:
        username: 'mqtt_username'
        password: 'mqtt_password'
    # Prefix to mqtt topic
    topicPrefix: 'knx'

```
The longer the bridge runs the more duplicates I get. I tried to increase the minimumDelay to more than 1s without a difference.

My KNX packet sniffer does not show these duplicates

Any idea why?

**Comment by giggo** [comment]
Here is a fix:

```
root@raspberrypi:/home/pi# diff knx-mqtt.js.orig knx-mqtt.js.geht.20190212
120a121,122
> var lastError = 0;
>
125c127,128
<         for (let key in groupAddresses) {
---
>         if (lastError == 0) {
>           for (let key in groupAddresses) {
134a138,141
>        }
>        else {
>       lastError = 0;
>        }
139a147,150
>       },
>       error: function(connstatus) {
>         logger.info("**** KNX ERROR: %j", connstatus);
>         lastError = 1;
141c152
<   }}, config.knx.options));
---
>   }}, config.knx));
```

**Comment by pakerfeldt** [comment]
I've been running my instances for many months and I'm unable to see the same issue as you. I'm a bit hesitant to merge that diff without knowing in more detail how it's solving the problem.

**Comment by shurli** [comment]
Same issue in my installation.
after same time messages are doubled.
even longer messages gets trippled and so on.

I'm very interessed in a fix.
I will test @giggo solution and will report my experience here.

**Comment by shurli** [comment]
it seems that knx connection is sometimes disconnected and reconnected.
On every connect event the eventhandler are add again.

The solution should be add the eventhandler only once, or remove them on "disconnect" event so it is ok to add them again on re-connect.

@giggo uses the `if (lastError == 0)` to prevent a mulitple add of eventhandlers.

**Comment by shurli** [comment]
I've implemented @giggo fix today.
There are several things to keep in mind:

.) line numbers of the diff doesn't match to the current master branch file.
the posted diff above is from feb19 and there was changes on master in april19

.) the last diff see below breaks the bridge
```
141c152
<   }}, config.knx.options));
---
>   }}, config.knx));
```
I had to ignore that and keep the argument
`config.knx.options`

_starting the long time test from now._

**Comment by pakerfeldt** [comment]
I would be happy to review a PR if you want to submit one. In case I would appreciate an explanation of why the fix solves the problem. Makes reviewing much easier.

**Comment by pakerfeldt** [comment]
Fixed in 0.0.22. See #8
